### PR TITLE
WIP euclidean distance algorithm server

### PR DESCRIPTION
**Pull request by raniarusly**
created RecommenderServlet:
- implemented overall class structure
- implement fetching recommendation rating
- implement distancePair class to represent a pair of recommendation ID and the euclidean distance to the user preference
- implement function to calculate distance between rating
- sort the top recommendation based on the distance
TODO: 
- fetch top 5 recommendation based on the calculated distance to send to the client
- implement client side

Lorna
NOTE: The majority of commits were incorrectly added (they belonged to other, irrelevant branches) but we can't work out how to remove them - any ideas? The only commits needed to be reviewed are @raniarusly commits and then the "Remove duplicate code".

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 1]
```diff
@@ -24,28 +23,53 @@
...
   }
 
-  // construct a matrix of size FACTORS_N x 1 which represents user's points
-  // every coloumn represents factor considered
-  private Matrix preferenceMatrix(HttpServletRequest request){
-    Matrix preference = new Matrix(FACTORS_N, 1);
-    for(int i = 0;  i < FACTORS_N; i++) {
-      int point = Integer.parseInt(request.getParameter(factors[i]));
-      preference.setValue(point, i, 1);
+  private List<Entity> getRecommendationsByCategory(String category){
```
Comment (lornaarmstrong) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]: This is duplicate code (the function exists just above it) so I'll remove it when I add my code to the algorithm.

Comment (raniarusly) [reply src/main/java/com/google/sps/servlets/RecommenderServlet.java]: coolm thanks Lorna

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 2]
```diff
@@ -14,7 +14,6 @@
 
 package com.google.sps.servlets;
 
-import com.google.sps.data.Matrix;
 import java.io.IOException;
```
Comment (lornaarmstrong) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java:26]: nit: there are missing import statements here (you use Pair, List etc.)

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 3]
```diff
@@ -24,28 +23,53 @@
...
-    // TODO: construct matrix of which represents the recommendations' points for every factor
-    return new Matrix(0,0);
+  private List<Pair<Integer, Integer>> getRecommendationsDistances(HttpServletRequest request){
+    String category = request.getParameter("category");
+    List<Entity> recommendations = getRecommendationsByCategory(category);
+
+    List<Pair<Integer, Integer>> distances = new ArrayList<>();
+    for(Entity recommendation : recommendations) {
+      int id= entity.getKey().getId();
+      int distance = calculateDistance(request, entity),
```
Comment (lornaarmstrong) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]:
```suggestion
      int distance = calculateDistance(request, entity);
```

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 4]
```diff
@@ -24,28 +23,53 @@
...
-  private Matrix recommendationMatix(){
-    // TODO: construct matrix of which represents the recommendations' points for every factor
-    return new Matrix(0,0);
+  private List<Pair<Integer, Integer>> getRecommendationsDistances(HttpServletRequest request){
+    String category = request.getParameter("category");
+    List<Entity> recommendations = getRecommendationsByCategory(category);
+
+    List<Pair<Integer, Integer>> distances = new ArrayList<>();
+    for(Entity recommendation : recommendations) {
+      int id= entity.getKey().getId();
```
Comment (lornaarmstrong) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]:
```suggestion
      int id = entity.getKey().getId();
```

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 5]
```diff
@@ -24,28 +23,53 @@
...
+    for(String factor : factors) {
+      query.addProjection(new PropertyProjection(factor, Integer.class));
+    }
+    return datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+  }
+
+  private Integer calculateDistance(HttpServletRequest request, Entity recommendation) {
+    int distance = 0;
+    for(String factor : factors) {
+      int recommendationValue = (int) recommendation.getProperty(factor);
```
Comment (lornaarmstrong) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]: I think either renaming variables or including comments would aid understanding here.

**Review by lornaarmstrong — COMMENTED** [review]
Overall, looks good - there are a few nits that need addressing.

Adding a couple of comments might make the code easier to understand, as it did take me a while to understand the names of variables etc.

**Review by lornaarmstrong — APPROVED** [review]
LGTM Thanks Rania!

File: src/main/java/com/google/sps/servlets/RecommenderServlet.java [hunk 6]
```diff
@@ -71,9 +71,10 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   private int calculateDistance(HttpServletRequest request, Entity recommendation) {
     int distance = 0;
     for(String factor : factors) {
-      int recommendationValue = (int) recommendation.getProperty(factor);
-      int preferenceValue = Integer.parseInt(request.getParameter(factor));
-      distance += Math.pow((recommendationValue - preferenceValue), 2);
+      //
```
Comment (lornaarmstrong) [inline src/main/java/com/google/sps/servlets/RecommenderServlet.java]: nit: please finish the comment here

**Comment by lornaarmstrong** [comment]
NOTE: The majority of commits were incorrectly added (they belonged to other, irrelevant branches) but we can't work out how to remove them - any ideas? The only commits needed to be reviewed are @raniarusly commits and then the "Remove duplicate code".

**Comment by raniarusly** [comment]
Rebasing problem

**Pull request closed without merging**